cli/parse.go

```diff
@@ -341,29 +265,142 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmdkit.Argument, recur
 				argDef.Required && !fillingVariadic {
 				r, err := maybeWrapStdin(stdin, msgStdinInfo)
 				if err != nil {
-					return nil, nil, err
+					return err
 				}
 
 				fpath := stdin.Name()
 				fileArgs[fpath] = files.NewReaderFile("", fpath, r, nil)
 			}
 		}
 
-		argDefIndex++
+		iArgDef++
+	}
+
+	if iArgDef == len(argDefs)-1 && stdin != nil &&
+		req.Command.Arguments[iArgDef].SupportsStdin {
+		// handle this one at runtime, pretend it's there
+		iArgDef++
 	}
 
 	// check to make sure we didn't miss any required arguments
-	if len(argDefs) > argDefIndex {
-		for _, argDef := range argDefs[argDefIndex:] {
+	if len(argDefs) > iArgDef {
+		for _, argDef := range argDefs[iArgDef:] {
 			if argDef.Required {
-				return nil, nil, fmt.Errorf("Argument '%s' is required", argDef.Name)
+				return fmt.Errorf("argument %q is required", argDef.Name)
+			}
+		}
+	}
+
+	req.Arguments = stringArgs
+	if len(fileArgs) > 0 {
+		req.Files = files.NewSliceFile("", "", filesMapToSortedArr(fileArgs))
```

Comment:
map to sorted array?

Reply:
I'm pretty sure I didn't come up with this and took it from the existing code. Probably from [here](https://github.com/ipfs/go-ipfs-cmds/pull/49/files/e58b7d68f5b9fa926fbe82eb9d037be7e5c34e51#diff-3be2b56996dc8a3b09575e16f3cc9ef0L364)